doc/design/figures/xgboost_feature_column.md

```diff
@@ -0,0 +1,131 @@
+
+## Feature column design in XGBoost models
+
+The training process of XGBoost model inside SQLFlow are as follows:
+
+- Step 1: Read data from database. Call the `db.db_generator()` method to return a Python generator which yields each row in database. 
+- Step 2: Dump SVM file. Call the `dump_dmatrix()` method to write the raw data into an SVM file. This file is ready to be loaded as XGBoost DMatrix.
+- Step 3: Training. Load the dumped file to be XGBoost DMatrix and start to train.
+
+Between step 1 and 2, we can do some pre-processing to the raw data from database. In Python side, we can implement the feature column interfaces like:
+
+```Python
+class BaseFeatureColumnTransformer(object):
+    def __call__(self, inputs):
+        raise NotImplementedError()
+
+class NumericColumnTransformer(BaseFeatureColumnTransformer):
+    # `column_idx` is the feature index inside `SELECT` statement
+    def __init__(self, column_idx):
+        self.column_idx = column_idx
+    
+    # `inputs` are all raw column data
+    # NumericColumnTransformer would only take the column indicated by `column_idx`
+    def __call__(self, inputs):
+        return inputs[self.column_idx]
+    
+# CategoryColumnTransformer is the base class of all category columns
+# This base class is design to do some check. For example, `INDICATOR`
+# would only accept category column as its input.
+class CategoryColumnTransformer(BaseFeatureColumnTransformer): pass
+        
+class BucketizedColumnTransformer(CategoryColumnTransformer):
+    def __init__(self, column_idx, boundaries, default_value=None):
+        self.column_idx = column_idx
+        self.boundaries = boundaries
+        self.default_value = default_value
+        
+    def __call__(self, inputs):
+        input = inputs[self.column_idx]
+        if input < boundaries[0]:
+            return 0
+            
+        for idx, b in enumarate(boudaries):
+            if input >= b
+                return idx
+        return len(boundaries)
+
+# CrossedColumnTransformer is designed as a decorator of CategoryColumnTransformer
+# It accepts several CategoryColumnTransformers, and crosses their transformed
+# results using hash algorithm.
+class CrossedColumnTransformer(CategoryColumnTransformer):
+    def __init__(self, *category_transformers):
+        for transformer in category_transformers:
+            assert isinstance(transformer, CategoryColumnTransformer)
+        self.category_transformers = category_transformers
+        
+    def _cross(self, transformed_inputs):
+        pass
+
+    def __call__(self, inputs):
+        transformed_inputs = []
+        for transformer in self.category_transformers:
+            transformed_inputs.append(transformer(inputs))
+         
+        self._cross(transformed_inputs)
+        
+class ListedFeatureColumnTransformer(BaseFeatureColumnTransformer):
```

Comment:
Maybe `ListedFeatureColumnTransformer` is not needed, we can call the transformers in a function call manner like: `IndicatorColumnTransformer(CategoryColumnWithHashBucketTransformer(...))`

Reply:
In this design, any transformer would only handle one `COLUMN` clause, and each clause would produce a list of transformer calls. `ListedFeatureColumnTransformer` is used to combine all `COLUMN` clauses. Therefore, maybe `ListedFeatureColumnTransformer` is needed to combine each column output.